scripts/generate-release-changelog-links.hs

```diff
@@ -0,0 +1,246 @@
+
+data CHaPEntry =
+  CHaPEntry { entryGitHubOwnerRepo :: Text
+            , entryGitHubRevision :: Text
+            , entrySubdir :: Maybe Text
+            }
+  deriving (Show)
+
+instance DecodeTOML CHaPEntry where
+  tomlDecoder =
+    CHaPEntry <$> getFields ["github", "repo"]
+              <*> getFields ["github", "rev"]
+              <*> optional (getField "subdir")
+
+findChangelogFromGitHub :: MonadIO m => CHaPEntry -> m (Maybe (Text, Text))
+findChangelogFromGitHub CHaPEntry{..} = do
+  (owner, repo) <- case splitOwnerRepo entryGitHubOwnerRepo of
+    Nothing -> die $
+      "Unable to split package owner and repository (" <> repr entryGitHubOwnerRepo <> ")"
+    Just (otxt, rtxt) -> pure (GitHub.mkName Proxy otxt, GitHub.mkName Proxy rtxt)
+
+  contentDir <- liftIO (runGitHub (changelogLookupGitHub owner repo entrySubdir entryGitHubRevision)) >>= \case
+    Left gitHubError -> die $
+      "GitHub lookup failed with error " <> repr gitHubError
+    Right (GitHub.ContentFile _) -> die $
+      "Expected changelogLookupGitHub to return a directory, but got a single file"
+    Right (GitHub.ContentDirectory dir) -> pure dir
+
+  pure $ case Data.Foldable.find looksLikeChangelog contentDir of
+    Nothing -> Nothing
+    Just res -> do
+      let name = GitHub.contentName (GitHub.contentItemInfo res)
+          path = GitHub.contentPath (GitHub.contentItemInfo res)
+      Just (name, constructGitHubPath owner repo entryGitHubRevision path)
+
+splitOwnerRepo :: Text -> Maybe (Text, Text)
+splitOwnerRepo ownerRepo =
+  case Text.breakOn "/" ownerRepo of
+    (_, "") -> Nothing
+    (pref, suff) -> Just (pref, Text.drop 1 suff)
+
+changelogLookupGitHub :: GitHub.Name GitHub.Owner
+                      -> GitHub.Name GitHub.Repo
+                      -> Maybe Text
+                      -> Text
+                      -> GitHub.Request k GitHub.Content
+changelogLookupGitHub owner repo subdir revision =
+  GitHub.contentsForR owner repo (fromMaybe "" subdir) (Just revision)
+
+looksLikeChangelog :: GitHub.ContentItem -> Bool
+looksLikeChangelog GitHub.ContentItem{..} = do
+  let caseInsensitiveName = CI.mk (GitHub.contentName contentItemInfo)
+  all id
+    [ contentItemType == GitHub.ItemFile
+    , caseInsensitiveName == "CHANGELOG.md"
+    ]
+
+constructGitHubPath :: GitHub.Name GitHub.Owner
+                    -> GitHub.Name GitHub.Repo
+                    -> Text
+                    -> Text
+                    -> Text
+constructGitHubPath owner repo revision path =
+  format ("https://github.com/"%ghname%"/"%ghname%"/blob/"%s%"/"%s)
+    owner repo revision path
+  where
+    ghname = makeFormat GitHub.untagName
+
+-- runGitHub :: GitHub.GitHubRO req res => req -> res
+-- runGitHub = GitHub.github'
+runGitHub :: GitHub.GitHubRW req res => req -> res
+runGitHub = GitHub.github (GitHub.OAuth "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")
```

Comment:
I'm not sure if this should be hardcoded in the code.

Reply:
ah, yeah, I was hitting rate limits when testing without an API key, but I meant to remove it (it's a temporary access key with no private permissions so not the end of the world though)